### ISSUE TITLE
Refactor tests to use data driven testing

### ISSUE THREAD

**Issue by hay-kot** [hhursev/recipe-scrapers#807]
Hello! 

I'm working on a project that involves recipe-scraper, but in another programming language. I'd like to build off the work that has been done here and keep my changes in sync with this library. 

As such, I had planned to extract the test cases in this library into a data driven model, where the HTML lives next to a JSON file with the expected structured data returned. This would allow me to write test in another language against that data.

That said, would this project be interested in me upstreaming those changes into this library? 

Selfishly it would be extremely helpful to me to have this library adopt a data driven model so I can easily stay up to date, but I completely understand if that's a huge refactor no one wants to review. 

--

If you're interested in this proposal I would suggest the following structure

```
/testdata
    /<domain>
      - index.html
      - expect.json
    /<domain>
      - index.html
      - expect.json
```

Then you would have one test method that would load that data dynamically for each directory and run the test against it. But I'm of course open to any suggestions on how you'd like this to be implemented.

**Comment by jayaddison** [comment]
Hey @hay-kot!  A relatively short answer from me while I think about it further: I _cautiously_ like this idea, because making the test expectations more like data could be powerful.

My concerns are:

  * Let's make sure we can support more than one test case per domain.
  * Are we sure JSON is the best format here?  Why not a CSV file for each test case (multiple rows for ingredients, instructions, other fields mostly singleton), or maybe Markdown or similar?
  * Would the results become more difficult for developers to edit and manage than the Python test modules we have today?  They're not easy to query/manage programatically, but they are simple for human editors, and that's good.

**Comment by hay-kot** [comment]
Thanks for taking a look!

I whipped up a proof of concept just to better flesh out some ideas. 

- https://github.com/hhursev/recipe-scrapers/pull/817

I think the implementation would be a little cleaner with pytest, but I didn't want to add a dependency if we didn't need to.

> Let's make sure we can support more than one test case per domain.

It will generate data driven tests for the base class any matching *.html and *.json file in a sub-directory in testdata. This should allow for multiple cases per domain. We can come up with some kind of convention to keep these clean.

> Are we sure JSON is the best format here? Why not a CSV file for each test case (multiple rows for ingredients, instructions, other fields mostly singleton), or maybe Markdown or similar?

I'd strongly prefer JSON because it is a _very_ standard format and nearly every single language has a built in parser for it. I'd also be okay with TOML or YAML as an alternative, but that would require an additional dev dependency.

I don't think markdown would be suitable because we have to build a custom parser, or rely on from matter for structured data, which at that point, it would probably make more sense to use YAML anyways.

CSVs are interesting. I not sure that editing that would be easier. I'd much prefer to edit a JSON/YAML/TOML file than a CSV file. Especially with any complex data types.

> Would the results become more difficult for developers to edit and manage than the Python test modules we have today? They're not easy to query/manage programmatically, but they are simple for human editors, and that's good.

IMO the current implementation is slightly confusing at first. I had a hard time grocking it initially because of 1) my unfamiliarity with unittest library and 2) the request mocking. My POC implementation relies on the `scrape_html` method so it would reduce some of the complexity there. 

Although, it does build tests dynamically at runtime, so it may be equally as confusing to get started with. 

Another benefit is we'd likely be able to consolidate all test code into a single file or two which would greatly reduce the cost of future refactors for testing.

**Comment by strangetom** [comment]
This is an interesting idea and I can see some advantages in this approach. 

One advantage would be we can make the tests more consistent across all the scrapers. At the moment, the tests implemented for each scraper tend to cover the same main things but other things aren't always tested, such as `canonical_url`. With this approach, it would be easier to autogenerate the file for the expected results for a new scraper and have the developer just fill in the fields.

Another advantage would be that it does away with some of the complexity that was added by inheriting the `test_consistent_ingredients_lists` test from the `ScraperTest` base class. We could still have this test, we just wouldn't have to deal with the inheritance issues we came across.

I agree with @hhursev on the use of json. We can already return json from a scraper, and a don't think the data a scraper returns is uniform enough for a format like csv.

On a practical note, it seems like we should be able to automatically convert the current test suite to this approach automatically, using `scraper.to_json()` to generate the expected output from the .`testhtml` files we have. There are a few subtleties in how we match each `.testhtml` to the correct domain that I'm not sure about, but should be possible.

**Comment by jayaddison** [comment]
>> Are we sure JSON is the best format here? Why not a CSV file for each test case (multiple rows for ingredients, instructions, other fields mostly singleton), or maybe Markdown or similar?
>
> I'd strongly prefer JSON because it is a very standard format and nearly every single language has a built in parser for it. I'd also be okay with TOML or YAML as an alternative, but that would require an additional dev dependency.
> 
> I don't think markdown would be suitable because we have to build a custom parser, or rely on from matter for structured data, which at that point, it would probably make more sense to use YAML anyways.
> 
> CSVs are interesting. I not sure that editing that would be easier. I'd much prefer to edit a JSON/YAML/TOML file than a CSV file. Especially with any complex data types.

Ok, I think I'm convinced that JSON would be the way to go here.  Thank you :+1: 

Probably a dumb question, but: if a scraper does not implement a field (no `ratings` field, for example), then should the developer simply omit that field from the corresponding JSON expectation file?

**Comment by hhursev** [comment]
That's a good idea and I think we should implement and update the package. I wish we built it that way to begin with 😅 Seems like the correct approach to me.

As @strangetom mentioned, there's a high chance the `.to_json()` method to make things easy.

On @jayaddison 's question about the case when method is not implemented, I'd say we omit the field from the json, yep.

I'll be able to get my hands dirty with this too first weekend of September (2nd / 3rd).

If you feel like it you can go all-in and implement the big changes, refactor the whole testing dir etc. and submit PR. However, even if you don't, the proposed change is going to be done by end of September 🤞

> That said, would this project be interested in me upstreaming those changes into this library?

tldr; yes. absolutely

**Comment by jayaddison** [comment]
cc @jknndy your feedback would be useful here if you'd like to participate in the discussion and shape the design of the next generation of the tests - in particular I think you've got a lot of recen experience developing multiple tests per scraper (for ingredient groups, for example).

**Comment by jknndy** [comment]
Is there any specific aspects of the grouping you'd like me to outline or look into? 

I'd also be interested in participating in this refactoring as a whole if additional help is needed.

**Comment by jayaddison** [comment]
I think the main thing to check is: if the test codebase is updated to implement the JSON-based test file approach that @hay-kot outlined, do you see that being better than the current Python-file based approach?  And if you do see any significant problems, let us know so we can think about and address those.

**Comment by strangetom** [comment]
The following (messy) code, run from the recipe_scrapers directory, will create the structure of test data described by @hay-kot above. For each test case, a json file of the expected output is created and placed alongside the `.testhtml` file in the directory for that domain i.e. `testdata/<domain>/domain.json` and `testdata/<domain>/domain.testhtml`.

There may be some tweaks we want to make, but it's a starting point. Since this is something we would only have to do once, I didn't focus much on tidying this code up.

```python
import json
import os
import shutil
import unittest

from recipe_scrapers import scrape_html


def get_test_class(test_case):
    """Find the scraper class for this test case"""
    for test in test_case:
        # If the test case only has one test, it's probably the
        # test inherited from the base class, so skip it.
        if test.countTestCases() > 1:
            try:
                return test._tests[0].scraper_class
            except:
                return None


def get_test_file(test_case):
    """Find the test file for this test case"""
    # _test[0] is always the test inherited from the base class,
    # so choose _test[1] first
    file = test_case._tests[1]._tests[0].test_file_name
    ext = test_case._tests[1]._tests[0].test_file_extension

    if file is not None:
        # test_file_name is set explicitly
        return f"{file}.{ext}"
    else:
        # test_file_name if determined from class name
        class_ = get_test_class(test_case)
        file = class_.__name__.lower()
        return f"{file}.{ext}"


if __name__ == "__main__":
    # Create new folder for new tests
    if not os.path.isdir("testdata"):
        os.mkdir("testdata")

    # Find all existing tests
    loader = unittest.TestLoader()
    suite = loader.discover("tests")

    for test in suite:
        class_ = get_test_class(test)
        if class_ is None:
            # Not a recipe scraper, but another test
            continue
        if "host" not in dir(class_):
            # Not a recipe scraper, probably _schemaorg
            continue

        testfile = get_test_file(test)
        host = class_.host()

        # Create folder for json output is doesn't exist
        if not os.path.isdir(f"testdata/{host}"):
            os.mkdir(f"testdata/{host}")

        testhtml = f"tests/test_data/{testfile}"
        jsonfile = f"testdata/{host}/{testfile.replace('testhtml', 'json')}"

        try:
            # Load testfile
            with open(testhtml, "r") as f:
                html = f.read()

            # Scrape test file
            scraper = scrape_html(html, org_url=host)

            # Write Scraper.to_json() to .json file in host folder
            with open(jsonfile, "w") as f:
                json.dump(scraper.to_json(), f, indent=2)

            # Copy testhtml file to host direcotry too
            shutil.copyfile(testhtml, f"testdata/{host}/{testfile}")

        except Exception as e:
            print(f"Skipped {class_.__name__} due to error: {e.args}")
```

There are a couple of tests where I get exceptions, which I think is down to using `scrape_html`, but I've not investigated.
* Gousto
* GoustoJson
* KptnCook
* MarleySpoon
* Woolworths

**Comment by jayaddison** [comment]
> There are a couple of tests where I get exceptions, which I think is down to using `scrape_html`, but I've not investigated.
> 
>     * Gousto
> 
>     * GoustoJson
> 
>     * KptnCook
> 
>     * MarleySpoon
> 
>     * Woolworths

From memory, this is approximately -- perhaps exactly -- the set of scrapers that are omitted from the `v15` branch because they use more than one HTTP request (often a JSON API call, or a second HTML fetch).

**Comment by jayaddison** [comment]
@hay-kot I'm looking at some of the output from the script at the moment - I like it :)

An idea primarily because I'm looking at some `instructions`-related stuff at the moment: should we special-case omit either the `instructions` field or the `instructions_list` field from the JSON files?  The contents of those fields should be equivalent in each test case.

**Comment by hay-kot** [comment]
> should we special-case omit either the instructions field or the instructions_list field from the JSON files? The contents of those fields should be equivalent in each test case.

I would say in the case of omitting the field we should assert that the correct exception was thrown, I think it would be `RecipeScrapersExceptions`, but I'm not 100% sure. I'd rather check that _something_ happens than pass the test by default, at least if there's a reasonable way to do it.

**Comment by jayaddison** [comment]
> > should we special-case omit either the instructions field or the instructions_list field from the JSON files? The contents of those fields should be equivalent in each test case.
> 
> I would say in the case of omitting the field we should assert that the correct exception was thrown, I think it would be `RecipeScrapersExceptions`, but I'm not 100% sure. I'd rather check that _something_ happens than pass the test by default, at least if there's a reasonable way to do it.

@hay-kot Ah, I didn't explain my question very well: I meant that it could make sense to filter-out either `instructions` or `instructions_list` from the `testdata/*/*.json` files.  At the moment the output of `to_json` includes both, so we find:

```json
{
    "instructions": "do the thing\ndo the second thing",
    "instructions_list": ["do the thing", "do the second thing"],
    ...
}
```

...which seems unnecessary in the test data.

(what you say also makes sense though: testing for exception cases would be nice)

**Comment by hhursev** [comment]
Has anybody started work on this? I didn't manage to start it couple weeks ago as I intended. Shall I start work on this now or someone is on it and halfway ready already 😅 

@jayaddison I'd say we duplicate the `instructions` and `instructions_list` in the test data as of now. No need to save lines.

**Comment by jayaddison** [comment]
Yep, @strangetom has a prototype / work-in-progress at #817.  I'm OK with duplicating both those fields to begin with :+1: although I think we should clean it up soon after, because it'll become annoying during bugfixes/patches

**Comment by hay-kot** [comment]
I haven't had a chance to circle back on data creation, IIRC the next step would be to use the work @strangetom did above to generate test files and combine that with my WIP PR and see how broken everything is 😅 and tidy up the code.

**Comment by strangetom** [comment]
Hi guys

I've spent a bit of time recently looking at implementing this. My current progress is at https://github.com/strangetom/recipe-scrapers/tree/data-driven-tests-refactor (which may become at PR at some point). This has been a case of taking @hhursev's proof concept and expanding it, plus converting all the current tests to the json format. So far it looks promising.

So far I've got to the point where you can run `python -m unittest` on the repository and it will run the tests against all the testjson files it finds, with 296 passes and 1 failure.  The tests for each scraper only tests a fixed number of scraper methods, which leaves some methods untested for scrapers that implement them.

I have excluded about 39 scrapers at the moment (see `BAD_SCRAPERS_LIST` in `tests/__init__.py`) because they cause errors. The main cause of the errors is that the scrapers don't implement the functions being tests (mostly `.yields()` and `.total_time()`).

There are a couple of things I'm planning on implementing next:
1. Because not all scraper implement all possible scraper functions, I think it would be a good idea to split the tests into two types:
    1. **Mandatory tests**, aligned with the mandatory functions listed in `docs/in-depth-guide-scraper-functions.md`, which all scraper will implemented the functions for.
    2. **Optional tests**. We could identify them by checking for any additional keys that are present in the testjson and if we find the key, run the test.
2. For the scrapers that don't implement a function for one of the mandatory tests, we could remove that key from the testjson. The test would then check that an exception is raised.
3. There's some extra work to do in updating the `generate.py` script to generate a default testjson file for a new scraper.


On a side note, I've noticed we have two scrapers for 101cookbooks and WeightWatchers, but only one of them gets used because in each case the two scrapers set the same `.host()`.

**Comment by jayaddison** [comment]
Huh, nice find related to those duplicate scrapers :)  That should be a quick fixup!

I'm +1 on the proposal of having some fields considered required and some considered optional -- and skipping the latter when a given field doesn't appear in the testjson.

Although I'm mulling it over, I think I also like the idea of there only being a single `RecipeTestCase` class.  There shouldn't really be any reason for a test class to adapt to the scraper class -- in fact, it's probably usually an antipattern / code smell when that happens.

Is there anything we can do to help?  (looking forward to code reviewing)

**Comment by strangetom** [comment]
Thanks for the offer @jayaddison. This has actually been going better than I thought it would so I think I'll be ready to open a PR this weekend. 

There will be a couple of outstanding issues that I will need some help with. The main one is that I can't see a way to support the `test_multiple_instructions` function in the current `ScraperTest` class. With the data driven approach there isn't a way to skip it like we currently can be overriding it in the derived class.
